### PR TITLE
[12.0] [MIG] web domain field

### PR DESCRIPTION
**Pull request by NL66278**
Tested and working very well!

### CODE REVIEW

File: web_domain_field/README.rst
Comment (omalbastin) [inline web_domain_field/README.rst:65]: but the same case can be achieved by specifying domain in field definition. For the example you have given, it can be achieved by 
```
testfield = fields.Char()
product_id = fields.Many2one('product.product',domain="[('type', '=', 'product'), ('name', 'like', testfield)]")
```
Maybe I am missing the whole picture here.

Comment (NL66278) [reply web_domain_field/README.rst:65]: @omalbastin The point is that the domain field can change at runtime.

**Comment by CasVissers-360ERP** [comment]
@NL66278 
Tried to test it but get:
```
Uncaught TypeError: CreateListFromArrayLike called on non-object
https://local/web_domain_field/static/src/js/pyeval.js:156
Traceback:
TypeError: CreateListFromArrayLike called on non-object
    at https://local/web_domain_field/static/src/js/pyeval.js:156:40
    at Function._.each._.forEach (https://local/web/static/lib/underscore/underscore.js:145:9)
    at _.<computed> [as each] (https://local/web/static/lib/underscore/underscore.js:1484:34)
    at eval_domains (https://local/web_domain_field/static/src/js/pyeval.js:152:20)
    at Object.domain_field_pyeval [as eval] (https://local/web_domain_field/static/src/js/pyeval.js:192:20)
    at OdooClass.stringToArray (https://local/web/static/src/js/core/domain.js:371:24)
    at Class._getDomain (https://local/web/static/src/js/views/basic/basic_model.js:3194:41)
    at Class._search (https://local/web_m2x_options/static/src/js/form.js:117:38)
    at https://local/web/static/src/js/fields/relational_fields.js:261:46
    at Function._.each._.forEach (https://local/web/static/lib/underscore/underscore.js:145:9)
```

**Comment by NL66278** [comment]
@CasVissers-360ERP How did you test this?

**Comment by CasVissers-360ERP** [comment]
Tested while developing an RFC    
```

product_id_domain = fields.Char(compute="_compute_product_id_domain", readonly=True, store=False)

@api.depends('order_id', 'order_id.analytic_account_id')
def _compute_product_id_domain(self):
        for rec in self:
            rec.product_id_domain = json.dumps([('sale_ok', '=', True),
                                                '|', ('company_id', '=', False), ('company_id', '=', rec.company_id)])
            if rec.order_id.analytic_account_id and rec.order_id.analytic_account_id.group_id and\
                rec.order_id.analytic_account_id.group_id.only_show_subscription_products:
                rec.product_id_domain = json.dumps([('sale_ok', '=', True), ('recurring_invoice', '=', True),
                                                    '|', ('company_id', '=', False), ('company_id', '=', rec.company_id)])

<xpath expr="//field[@name='product_id']" position='before'>
    <field name="product_id_domain" invisible="1"/>
</xpath>

<xpath expr="//field[@name='product_id']" position='attributes'>
       <attribute name="domain">product_id_domain</attribute>
</xpath>
```
@NL66278 
EDIT: Looks like I forgot to add ID to rec.company_id.id, testing again.
EDIT 2: Still same error

**Comment by NL66278** [comment]
@CasVissers-360ERP I have added an web_domain_field_example module for easier test/demo. Please check it out.

Can I compare what you made somewhere? Is the code available.

**Comment by CasVissers-360ERP** [comment]
@NL66278 
Error on my side I forgot to add the field to the form view on sale order line form view which was loaded:

This fixed it:
```
            <xpath expr="//field[@name='order_line']/tree/field[@name='product_id']" position="before">
                <field name="product_id_domain" invisible="1"/>
            </xpath>

            <xpath expr="//field[@name='order_line']/form/group/group/field[@name='product_id']" position="before">
                <field name="product_id_domain" invisible="1"/>
            </xpath>

            <xpath expr="//field[@name='order_line']/tree/field[@name='product_id']" position="attributes">
                <attribute name="domain">product_id_domain</attribute>
            </xpath>

            <xpath expr="//field[@name='order_line']/form/group/group/field[@name='product_id']" position="attributes">
                <attribute name="domain">product_id_domain</attribute>
            </xpath>
```

**Review by CasVissers-360ERP — APPROVED** [review]

**Review by marcelsavegnago — APPROVED** [review]

**Comment by NL66278** [comment]
@lmignon Hi Laurence, can you have a look at this PR, and merge if you approve?

**Review by lmignon — APPROVED** [review]

**Comment by lmignon** [comment]
@NL66278 I'm not part of the Web PSC... therefore I can't merge this pr.

**Comment by NL66278** [comment]
@pedrobaeza @hbrunn @simahawk I need this for a much improved version of partner_multi_relation. Could one of you please have a look and merge if you approve (or suggest changes if otherwise)?

**Comment by NL66278** [comment]
@lmignon Thanks anyway for your approval!

**Comment by pedrobaeza** [comment]
I wonder if you can't do what you want with domains based on compute m2m fields. I haven't got the need of using this module yet.

**Comment by rousseldenis** [comment]
> I wonder if you can't do what you want with domains based on compute m2m fields. I haven't got the need of using this module yet.

I remember there was a limitation but maybe @lmignon can help to find out exactly ?

**Comment by NL66278** [comment]
A domain with a computed domain field would translate into a search using an array of id's. In our use case that could be tens of thousands of id's and therefore be unworkable. The web_domain_field solution works for our customers in practice and is quite efficient.

**Comment by lmignon** [comment]
Here it's the motivations behind this addons... https://github.com/odoo/odoo/pull/17328

**Comment by NL66278** [comment]
Any maintainer willing to merge this? I have another PR in partner-contact depending on this one.

**Review by CasVissers-360ERP — APPROVED** [review]

**Comment by NL66278** [comment]
@Tardo or @tarteo Would it be possible to merge this one?

**Comment by CharellKing** [comment]
in odoo14, i have the same error:
Traceback:
TypeError: CreateListFromArrayLike called on non-object
    at http://10.227.12.242:18469/web/content/167-d6938d5/web.assets_backend.js:360:20
    at Function._.each._.forEach (http://10.227.12.242:18469/web/content/102-1162942/web.assets_common.js:108:558)
    at _.<computed> [as each] (http://10.227.12.242:18469/web/content/102-1162942/web.assets_common.js:165:526)
    at eval_domains (http://10.227.12.242:18469/web/content/167-d6938d5/web.assets_backend.js:357:265)
    at Object.pyeval [as eval] (http://10.227.12.242:18469/web/content/167-d6938d5/web.assets_backend.js:382:24)
    at OdooClass.stringToArray (http://10.227.12.242:18469/web/content/167-d6938d5/web.assets_backend.js:324:674)
    at Class._getDomain (http://10.227.12.242:18469/web/content/167-d6938d5/web.assets_backend.js:1504:167)
    at Class._search (http://10.227.12.242:18469/web/content/167-d6938d5/web.assets_backend.js:1097:607)
    at http://10.227.12.242:18469/web/content/167-d6938d5/web.assets_backend.js:1089:260
    at Function._.each._.forEach (http://10.227.12.242:18469/web/content/102-1162942/web.assets_common.js:108:558)


`class InspectOrder(models.Model):
    _inherit = "sale.order"

    provider_id_domain = fields.Char(compute="_computer_provider_id_domain", readonly=True, store=False)
    provider_id = fields.Many2one("res.partner", string=_("Provider"), readonly=True,
                                  states={'draft': [('readonly', False)], 'sent': [('readonly', False)]}, required=True,
                                  index=True, tracking=1)
    
    partner_id_domain = fields.Char(compute="_computer_partner_id_domain", readonly=True, store=False)

    
    def _filter_companies(self, *args):
        customer_companies = self.env['res.partner'].search(
            ['&', ('is_company', '=', True), '&',
             ('id', 'not in', args),
             ('id', '!=', self.env.user.partner_id.id)])
        
        companies_ids = []
        for customer_company in customer_companies:
            companies_ids.append(customer_company.id)
        return companies_ids
    
    @api.depends("partner_id")
    def _computer_provider_id_domain(self):
        for rec in self:
            rec.provider_id_domain = json.dumps(
                [('id', 'in', self._filter_companies(self.partner_id.id))]
            )
            
    @api.depends("provider_id")
    def _computer_partner_id_domain(self):
        for rec in self:
            rec.partner_id_domain = json.dumps(
                [('id', 'in', self._filter_companies(self.provider_id.id))]
            )
`

`
<odoo>
    <data>
        <record id="trade_party2nd_inspect_order" model="ir.ui.view">
            <field name="name">trade_party2nd_inspect_order</field>
            <field name="model">sale.order</field>
            <field name="inherit_id" ref="sale.view_order_form"/>
                <field name="arch" type="xml">
                <xpath expr="//field[@name='partner_id']" position="before">
                    <field name="partner_id_domain" invisible="1"/>
                    <field name="provider_id_domain" invisible="1"/>
                </xpath>
                <xpath expr="//field[@name='partner_id']" position="attributes">
                    <attribute name="domain">partner_id_domain</attribute>
                </xpath>
                <xpath expr="//field[@name='partner_id']" position="after">
                    <field name="customer_contacts_ids"/>
                </xpath>
            </field>
        </record>
    </data>
</odoo>

`

**Comment by quzhaojing** [comment]
in odoo14, i have the same error:
Traceback:
TypeError: CreateListFromArrayLike called on non-object
    at /web/content/167-d6938d5/web.assets_backend.js:360:20
    at Function._.each._.forEach (/web/content/102-1162942/web.assets_common.js:108:558)
    at _.<computed> [as each] (/web/content/102-1162942/web.assets_common.js:165:526)
    at eval_domains (/web/content/167-d6938d5/web.assets_backend.js:357:265)
    at Object.pyeval [as eval] (/web/content/167-d6938d5/web.assets_backend.js:382:24)
    at OdooClass.stringToArray (/web/content/167-d6938d5/web.assets_backend.js:324:674)
    at Class._getDomain (/web/content/167-d6938d5/web.assets_backend.js:1504:167)
    at Class._search (/web/content/167-d6938d5/web.assets_backend.js:1097:607)
    at /web/content/167-d6938d5/web.assets_backend.js:1089:260
    at Function._.each._.forEach (/web/content/102-1162942/web.assets_common.js:108:558)


```python
class InspectOrder(models.Model):
    _inherit = "sale.order"

    provider_id_domain = fields.Char(compute="_computer_provider_id_domain", readonly=True, store=False)
    provider_id = fields.Many2one("res.partner", string=_("Provider"), readonly=True,
                                  states={'draft': [('readonly', False)], 'sent': [('readonly', False)]}, required=True,
                                  index=True, tracking=1)
    
    partner_id_domain = fields.Char(compute="_computer_partner_id_domain", readonly=True, store=False)

    
    def _filter_companies(self, *args):
        customer_companies = self.env['res.partner'].search(
            ['&', ('is_company', '=', True), '&',
             ('id', 'not in', args),
             ('id', '!=', self.env.user.partner_id.id)])
        
        companies_ids = []
        for customer_company in customer_companies:
            companies_ids.append(customer_company.id)
        return companies_ids
    
    @api.depends("partner_id")
    def _computer_provider_id_domain(self):
        for rec in self:
            rec.provider_id_domain = json.dumps(
                [('id', 'in', self._filter_companies(self.partner_id.id))]
            )
            
    @api.depends("provider_id")
    def _computer_partner_id_domain(self):
        for rec in self:
            rec.partner_id_domain = json.dumps(
                [('id', 'in', self._filter_companies(self.provider_id.id))]
            )
```

```xml
<odoo>
    <data>
        <record id="trade_party2nd_inspect_order" model="ir.ui.view">
            <field name="name">trade_party2nd_inspect_order</field>
            <field name="model">sale.order</field>
            <field name="inherit_id" ref="sale.view_order_form"/>
                <field name="arch" type="xml">
                <xpath expr="//field[@name='partner_id']" position="before">
                    <field name="partner_id_domain" invisible="1"/>
                    <field name="provider_id_domain" invisible="1"/>
                </xpath>
                <xpath expr="//field[@name='partner_id']" position="attributes">
                    <attribute name="domain">partner_id_domain</attribute>
                </xpath>
                <xpath expr="//field[@name='partner_id']" position="after">
                    <field name="provider_id" domain="provider_id_domain"/>
                </xpath>
            </field>
        </record>
    </data>
</odoo>
```

**Comment by lmignon** [comment]
@CharellKing 

> ```python
> class InspectOrder(models.Model):
>     _inherit = "sale.order"
> 
>     provider_id_domain = fields.Char(compute="_computer_provider_id_domain", readonly=True, store=False)
>     provider_id = fields.Many2one("res.partner", string=_("Provider"), readonly=True,
>                                   states={'draft': [('readonly', False)], 'sent': [('readonly', False)]}, required=True,
>                                   index=True, tracking=1)
>     
>     partner_id_domain = fields.Char(compute="_computer_partner_id_domain", readonly=True, store=False)
> 
>     
>     def _filter_companies(self, *args):
>         customer_companies = self.env['res.partner'].search(
>             ['&', ('is_company', '=', True), '&',
>              ('id', 'not in', args),
>              ('id', '!=', self.env.user.partner_id.id)])
>         
>         companies_ids = []
>         for customer_company in customer_companies:
>             companies_ids.append(customer_company.id)
>         return companies_ids
>     
>     @api.depends("partner_id")
>     def _computer_provider_id_domain(self):
>         for rec in self:
>             rec.provider_id_domain = json.dumps(
>                 [('id', 'in', self._filter_companies(self.partner_id.id))]
>             )
>             
>     @api.depends("provider_id")
>     def _computer_partner_id_domain(self):
>         for rec in self:
>             rec.partner_id_domain = json.dumps(
>                 [('id', 'in', self._filter_companies(self.provider_id.id))]
>             )
> ```

Your computed domain could be wrong in the case where  `self._filter_companies` return an empty list. 

```python

import odoo.osv.expression as expression

...

@api.depends("provider_id")
def _compute_partner_id_domain(self):
    for rec in self:
        alllowed_partner_ids = self._filter_companies(self.provider_id.id)
        if allowed_partner_ids:
            domain = [('id', 'in', allowed_partner_ids)]]
        else:
            domain = expression.FALSE_DOMAIN
        rec.partner_id_domain = json.dumps(domain) 
```

 >     provider_id_domain = fields.Char(compute="_computer_provider_id_domain", readonly=True, store=False)

FYI  computed fields are read only and not stored by default...
```python
 provider_id_domain = fields.Char(compute="_computer_provider_id_domain")
```  

>     def _filter_companies(self, *args):
>         customer_companies = self.env['res.partner'].search(
>             ['&', ('is_company', '=', True), '&',
>              ('id', 'not in', args),
>              ('id', '!=', self.env.user.partner_id.id)])
>         
>         companies_ids = []
>         for customer_company in customer_companies:
>             companies_ids.append(customer_company.id)
>         return companies_ids

The result of a call to `search` is a recordset. You can get the list of all the ids into the recordset bu using the property `ids` of the recordset. 

```python
     def _filter_companies(self, partner_id):
         return self.env['res.partner'].search(
             [('is_company', '=', True),
              ('id', '!=', partner_id),
              ('id', '!=', self.env.user.partner_id.id)]).ids
```

**Comment by quzhaojing** [comment]
it's has values;  value is "[[\"id\", \"in\", [1, 14, 18]]]"

result_domain.push.apply(result_domain,domain_array_to_combine)

domain_array_to_combine is "[[\"id\", \"in\", [1, 14, 18]]]"

in javascript push.apply's second params must be array, not a string;

**Comment by quzhaojing** [comment]
almost 3 years ago, someone has open such issue, but not solve: https://github.com/OCA/web/issues/882

**Comment by lmignon** [comment]
@CharellKing This addon is not yet migrated in odoo14?!

**Comment by lmignon** [comment]
@CharellKing If your problem is with the code from the migration PR, you should write your comment into this PR https://github.com/OCA/web/pull/1778 . The current PR  is not for odoo 14

**Comment by quzhaojing** [comment]
> @CharellKing If your problem is with the code from the migration PR, you should write your comment into this PR #1778 . The current PR is not for odoo 14

OK, thank you

**Comment by NL66278** [comment]
@dreispt Thanks a lot for finally merging this one!

**Comment by 17621706783** [comment]
Does it support odoo11?

**Comment by NL66278** [comment]
@17621706783 This is specifically for Odoo 12.0. As a lot has changed in the javascript functionality in Odoo, it should not be assumed that this module can be used on 11.0. The module is available in 10.0, 13.0 and 14.0.

**Pull request merged**
